third_party/tlx/tutorials/hopper-persistent-gemm-ws-pingpong,py

```diff
@@ -0,0 +1,314 @@
+    b = tlx.local_alloc((BK, BN), tlx.dtype_of(b_desc), NUM_STAGES)
+
+    # Mainloop Barriers: For producer-consumer synchronization on A and B buffers.
+    # The producer waits on empty, consumers wait on full.
+    mainloop_empty_bar = tlx.alloc_barriers(num_barriers=NUM_STAGES, arrive_count=1)
+    mainloop_full_bar = tlx.alloc_barriers(num_barriers=NUM_STAGES, arrive_count=1)
+
+    pingpong_mma_bar = tlx.alloc_barriers(num_barriers=1, arrive_count=1)
+    pingpong_epi_bar = tlx.alloc_barriers(num_barriers=1, arrive_count=1)
+    with tlx.async_tasks():
+        # Producer (async load)
+        with tlx.async_task("default"):
+            start_pid = tl.program_id(axis=0)
+            num_pid_m = tl.cdiv(M, BM)
+            num_pid_n = tl.cdiv(N, BN)
+            num_tiles = num_pid_m * num_pid_n
+            num_pid_in_group = GROUP_SIZE_M * num_pid_n
+            
+            p = 1
+            buf = 0
+            
+            for tile_id in range(start_pid, num_tiles, NUM_SMS):
+                group_id = tile_id // num_pid_in_group
+                first_pid_m = group_id * GROUP_SIZE_M
+                group_size_m = min(num_pid_m - first_pid_m, GROUP_SIZE_M)
+                pid_m = first_pid_m + ((tile_id % num_pid_in_group) % group_size_m)
+                pid_n = (tile_id % num_pid_in_group) // group_size_m
+                
+                offset_am = pid_m * BM
+                offset_bn = pid_n * BN
+
+                for k in range(0, tl.cdiv(K, BK)):
+                    offset_k = k * BK
+                    
+                    # Async load to a[buf] and b[buf]
+                    empty = tlx.local_view(mainloop_empty_bar, buf)
+                    full = tlx.local_view(mainloop_full_bar, buf)
+                    tlx.barrier_wait(bar=empty, phase=p)
+                    tlx.barrier_expect_bytes(full, BM * BK * 2 + BK * BN * 2) # a and b
+                    data_a = tlx.local_view(a, buf)
+                    tlx.async_descriptor_load(a_desc, data_a, [offset_am, offset_k], full)
+                    data_b = tlx.local_view(b, buf)
+                    tlx.async_descriptor_load(b_desc, data_b, [offset_k, offset_bn], full)
+
+                    p = p ^ (buf == (NUM_STAGES - 1))
+                    buf = (buf + 1) % NUM_STAGES
+
+        # Consumers (wgmma + async store)
+        with tlx.async_task(num_warps=4, replicate=2, registers=232):
+            cid: tl.constexpr = tlx.async_task_replica_id()
+            start_pid = tl.program_id(axis=0) + cid * NUM_SMS
+            num_pid_m = tl.cdiv(M, BM)
+            num_pid_n = tl.cdiv(N, BN)
+            num_tiles = num_pid_m * num_pid_n
+            num_pid_in_group = GROUP_SIZE_M * num_pid_n
+            k_tiles = tl.cdiv(K, BK)
+
+            tile_rank = cid # cta0: 0, 2, 4 cta1; 1, 3, 5
+            phase_mma = 1 - cid
+            phase_epi = 1 - cid
+
+            mma_bar = tlx.local_view(pingpong_mma_bar, 0)
```

Comment:
This implementation works pretty well. Just fyi there is another way to implement pingpong, i.e, with named barriers, which is sometimes slightly more efficient. For example, https://github.com/facebookexperimental/triton/blob/a6e1c5b0c34ff841aa2045c17a4090e80886df2d/third_party/tlx/tutorials/hopper-fa-ws-pipelined-pingpong_test.py#L135